### PR TITLE
Trial class average completion script

### PR DESCRIPTION
**Pull request by differentmatt**

### CODE REVIEW

File: scripts/analytics/trialClassAverageCompletion.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+// 4. Find first non-empty cs1 course instances for teachers
+// 5. Find per-classroom cs1 course completion averages
+// 6. Compute overall average
+
+// NOTE: update the endDay variable below to control the date range
+// NOTE: assumes a couple unchanged course and first level Ids
+
+// TODO: use lodash
+// TODO: better handling of archived classrooms
+// TODO: corner cases around classroom membership changes not reflected in course instances
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: What corner cases?

File: scripts/analytics/trialClassAverageCompletion.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+if (process.argv.length !== 4) {
+  console.log("Usage: node <script> <mongo connection Url> <mongo connection Url level session>");
+  console.log("Include ?readPreference=secondary in connection URLs");
+  process.exit();
+}
+
+'use strict';
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: `'use strict';` needs to be the first line in the file/code block. Move this to line 1.

File: scripts/analytics/trialClassAverageCompletion.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+const mongoConnUrl = process.argv[2];
+const mongoConnUrlLevelSessions = process.argv[3];
+
+const debugOutput = false;
+
+const cs1CampaignId = '55b29efd1cd6abe8ce07db0d';
+const cs1CourseId = '560f1a9f22961295f9427742';
+
+const daysToFirstLevelSession = 7;
+const daysToCompleteTrial = 7;
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: What is a trial in this context? Not mentioned in the comments at the top

File: scripts/analytics/trialClassAverageCompletion.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
+// Find average classroom completion rate for first week of cs1 activity
+
+// Algorithm
+// 1. Find initial classroom student activity for first cs1 level between start and midpoint times
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: What is this about a midpoint? Could use clarification

File: scripts/analytics/trialClassAverageCompletion.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+const endDay = "2017-01-26";
+let startDay = new Date(`${endDay}T00:00:00.000Z`);
+startDay.setUTCDate(startDay.getUTCDate() - daysToCompleteTrial);
+const midpointDay = startDay.toISOString().substring(0, 10);
+startDay.setUTCDate(startDay.getUTCDate() - daysToFirstLevelSession);
+startDay = startDay.toISOString().substring(0, 10);
+const startObjectId = objectIdWithTimestamp(new Date(`${startDay}T00:00:00.000Z`));
+const midpointObjectId = objectIdWithTimestamp(new Date(`${midpointDay}T00:00:00.000Z`));
+const endObjectId = objectIdWithTimestamp(new Date(`${endDay}T00:00:00.000Z`));
+console.log(`Measuring days ${startDay} to ${midpointDay} to ${endDay}`);
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: Script needs to error out if `new Date(`${endDay}T00:00:00.000Z`) > new Date()`. Time zones don't make it obvious when that's the case (it's at 4pm the day before)

File: scripts/analytics/trialClassAverageCompletion.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    if (!cs1Campaign.levels[originalId].practice) {
+      cs1OriginalLevelIds.push(originalId);
+    }
+  }
+  debug(`cs1 levels ${cs1OriginalLevelIds.length}`);
+
+  // Find initial classroom student activity for first cs1 level between start and midpoint times
+  console.log(`Finding initial activity between ${startDay} and ${midpointDay}..`);
+  const firstLevelSessions = yield lsDb.collection('level.sessions').find(
+    {$and: [{_id: {$gte: startObjectId}}, {_id: {$lt: midpointObjectId}}, {'level.original': cs1OriginalLevelIds[0]}, {heroConfig: {$exists: false}}]},
```
Comment (phoenixeliot) [inline scripts/analytics/trialClassAverageCompletion.js]: Why does this use `{heroConfig: {$exists: false}}`?

File: scripts/analytics/trialClassAverageCompletion.js (hunk 1 above)
Comment (differentmatt) [inline scripts/analytics/trialClassAverageCompletion.js]: Seems like classroom deleted members are not removed form course instances, but I did not investigate further.

File: scripts/analytics/trialClassAverageCompletion.js (hunk 3 above)
Comment (differentmatt) [inline scripts/analytics/trialClassAverageCompletion.js]: cs1 at moment, will rename variable.

File: scripts/analytics/trialClassAverageCompletion.js (hunk 4 above)
Comment (differentmatt) [inline scripts/analytics/trialClassAverageCompletion.js]: Good catch, will update comments.

File: scripts/analytics/trialClassAverageCompletion.js (hunk 5 above)
Comment (differentmatt) [inline scripts/analytics/trialClassAverageCompletion.js]: 👍

File: scripts/analytics/trialClassAverageCompletion.js (hunk 6 above)
Comment (differentmatt) [inline scripts/analytics/trialClassAverageCompletion.js]: Forgot we added a better flag in `isForClassroom`, will change.

**Comment by differentmatt** [comment]
Closed by typo'd commit https://github.com/codecombat/codecombat/commit/1308c23aa7dec28ef3331f55341d2b9e62ea88ed

**Pull request closed without merging**